iac-extensions/docker/src/main/java/org/sonar/iac/docker/checks/DirectoryCopySourceCheck.java

```diff
@@ -0,0 +1,119 @@
+    for (SyntaxToken src : add.srcs()) {
+      String path = SyntaxTokenUtils.sanitize(src);
+      if (!path.startsWith("http://") && !path.startsWith("https://")) {
+        reportIfSensitive(ctx, src, isSensitivePath(path), "ADD");
+      }
+    }
+  }
+
+  private static void checkCopy(CheckContext ctx, CopyTree copy) {
+    if (hasOption(copy.options(), "from")) return;
+
+    for (SyntaxToken src : copy.srcs()) {
+      reportIfSensitive(ctx, src, isSensitivePath(SyntaxTokenUtils.sanitize(src)), "COPY");
+    }
+  }
+
+  private static boolean hasOption(List<ParamTree> options, String key) {
+    return options.stream().anyMatch(param -> param.name().equals(key));
+  }
+
+  private static void reportIfSensitive(CheckContext ctx, SyntaxToken src, PathSensitivity sensitivity, String instructionName) {
+    if(sensitivity == PathSensitivity.ROOT_OR_CURRENT) {
+      ctx.reportIssue(src, MESSAGE_CURRENT_OR_ROOT);
+    } else if (sensitivity == PathSensitivity.TOP_LEVEL_GLOBBING) {
+      ctx.reportIssue(src, String.format(MESSAGE_GLOBING, instructionName));
+    }
+  }
+
+  enum PathSensitivity {
+    SAFE, ROOT_OR_CURRENT, TOP_LEVEL_GLOBBING
+  }
+
+  /**
+   * Indicate if a path is sensitive. A path is sensitive if any of the current condition is true :
+   * - if the provided path resolve to a root level or to the current directory
+   * - if there is any top level entry (level <= 1) ending with a wildcard *
+   * Examples of root level : '/' ; '/test/..' ; 'c:/' ; 'c:/test/..'
+   * Examples of current level : '.' ; './test/..'
+   * Examples of top level entry ending with wildcard : 'a*' ; './a*' ; '/a*' ; './test/../a*'
+   */
+  private static PathSensitivity isSensitivePath(String path) {
+    String[] splitted = path.split("/");
+    String regexWindowsDrive = "^[a-zA-Z]:$";
+    boolean isAbsolutePath = splitted.length == 0 || splitted[0].isEmpty() || splitted[0].matches(regexWindowsDrive);
+    int level = 0;
+    boolean hasTopLevelWithWildcard = false;
+
+    for (int i = 0; i < splitted.length; i++) {
+      if (splitted[i].equals("..")) {
+        level--;
+      } else if (!splitted[i].equals(".") && !splitted[i].isEmpty() && (i > 0 || !splitted[i].matches(regexWindowsDrive))) {
+        level++;
+      }
+
+      if (level <= 1 && splitted[i].endsWith("*")) {
+        hasTopLevelWithWildcard = true;
+      }
+
+      if (level < 0) {
+        if (isAbsolutePath) {
+          // in case of absolute path, we can't go above root, so we reset to root
+          level = 0;
+        } else {
+          // in case of relative path, if we go above the current folder, we suppose we lose the track and can't report any issue
+          return PathSensitivity.SAFE;
+        }
+      }
+    }
+
+    if (level == 0) return PathSensitivity.ROOT_OR_CURRENT;
+    if (level == 1 && hasTopLevelWithWildcard) return PathSensitivity.TOP_LEVEL_GLOBBING;
+    return PathSensitivity.SAFE;
```

Comment:
would it be possible to implement it using `Path`? I think some operations might be simplified, e.g. we might use `path.normalize()` to resolve the `..`

Reply:
I tried, Path is system dependent.

Although it give really great result especially from Unix system (still it may give strange result with driver letter format such as C:/), on windows we get blocked quite quickly as * is not an authorized character. We get also different result, so it implies that we support any potential result.

It is apparently not possible to force the Unix Path format, the class used in that case come directly from installed FileSystemProviders, which are defined in the JDK installed on the machine running the code.

I found some shady solutions online about building some memory file system, but it seems a bit too much for our usage :D